servers/graphql-kotlin-server/src/main/kotlin/com/expediagroup/graphql/server/execution/GraphQLRequestHandler.kt

```diff
@@ -16,37 +16,125 @@
+                        instrumentation.javaClass == DataLoaderSyncExecutionExhaustedInstrumentation::class.java
+                }
+                ?.javaClass
+        }
+
     /**
      * Execute a GraphQL request in a non-blocking fashion.
      * This should only be used for queries and mutations.
      * Subscriptions require more specific server logic and will need to be handled separately.
      */
-    open suspend fun executeRequest(request: GraphQLRequest, context: GraphQLContext? = null, graphQLContext: Map<*, Any> = emptyMap<Any, Any>()): GraphQLResponse<*> {
-        // We should generate a new registry for every request
-        val dataLoaderRegistry = dataLoaderRegistryFactory?.generate()
-        val executionInput = request.toExecutionInput(context, dataLoaderRegistry, graphQLContext)
+    open suspend fun executeRequest(
+        graphQLRequest: GraphQLServerRequest,
+        context: GraphQLContext? = null,
+        graphQLContext: Map<*, Any> = emptyMap<Any, Any>()
+    ): GraphQLServerResponse =
+        when (graphQLRequest) {
+            is GraphQLRequest -> {
+                execute(
+                    graphQLRequest.toExecutionInput(context, dataLoaderRegistryFactory?.generate(), graphQLContext)
+                )
+            }
+            is GraphQLBatchRequest -> {
+                when {
+                    graphQLRequest.requests.none(GraphQLRequest::isMutation) -> {
+                        val dataLoaderRegistry = dataLoaderRegistryFactory?.generate()
+                        val contextWithDataLoaderInstrumentationState = graphQLContext + (
+                            dataLoaderRegistry?.let {
+                                when (batchingInstrumentationStrategy) {
+                                    DataLoaderLevelDispatchedInstrumentation::javaClass -> mapOf(
+                                        ExecutionLevelDispatchedState::class to ExecutionLevelDispatchedState(
+                                            graphQLRequest.requests.size
+                                        )
+                                    )
+                                    DataLoaderSyncExecutionExhaustedInstrumentation::javaClass -> mapOf(
+                                        SyncExecutionExhaustedState::class to SyncExecutionExhaustedState(
+                                            graphQLRequest.requests.size,
+                                            dataLoaderRegistry
+                                        )
+                                    )
+                                    else -> null
+                                }
+                            } ?: emptyMap()
+                        )
+                        GraphQLBatchResponse(
+                            execute(
+                                graphQLRequest.requests.map {
+                                    it.toExecutionInput(context, null, contextWithDataLoaderInstrumentationState)
```

Comment:
if batching enabled either using by `level` or `sync execution exhaustion`, do not add the `DataLoaderRegistry` to the execution input, no need to, the `DataLoaderRegistry` is located in the shared `GraphQLContext`